File: leetcode_problems/typescript/src/lc2506_count_pairs_of_similar_strings.ts
function arraysEqual(arr1: boolean[], arr2: boolean[]): boolean {
  if (arr1.length !== arr2.length) return false;
  for (let i = 0; i < arr1.length; i++) {
    if (arr1[i] !== arr2[i]) return false;
  }
  return true;
}

function similarPairs(words: string[]): number {
  let count = 0;
  let word_chars: boolean[][] = [];
  for (let i = 0; i < words.length; i++) {
    word_chars.push(new Array(26).fill(false));
    let word = words[i];
    for (let char of word) {
      word_chars[i][char.charCodeAt(0) - "a".charCodeAt(0)] = true;
    }
  }

  for (let i = 0; i < word_chars.length; i++) {
    for (let j = i + 1; j < word_chars.length; j++) {
      if (arraysEqual(word_chars[i], word_chars[j])) {
        count++;
      }
    }
  }

  return count;
}

function similarPairs2(words: string[]): number {
  let count = 0;
  const word_masks = new Map<number, number>();
  for (let word of words) {
    let mask = 0;
    for (let char of word) {
      mask |= 1 << (char.charCodeAt(0) - "a".charCodeAt(0));
    }
    word_masks.set(mask, (word_masks.get(mask) || 0) + 1);
  }
  word_masks.forEach((v, k) => {
    count += (v * (v - 1)) / 2;
  });
  return count;
}

console.log(similarPairs2(["aba", "aabb", "abcd", "bac", "aabc"]));

File: leetcode_problems/cpp/lc2506_count_pairs_of_similar_strings.cpp
#include <string>
#include <vector>
#include <map>
#include <set>
#include <unordered_map>
#include <iostream>
using std::map;
using std::set;
using std::string;
using std::unordered_map;
using std::vector;

class Solution
{
public:
    int similarPairs(vector<string> &words)
    {
        int count = 0;
        vector<vector<bool>> word_chars(words.size(), vector<bool>(26, false));
        for (int i = 0; i < words.size(); i++)
        {
            string word = words[i];
            for (char chr : word)
            {
                word_chars[i][chr - 'a'] = true;
            }
        }

        for (int i = 0; i < word_chars.size(); i++)
        {
            for (int j = i + 1; j < word_chars.size(); j++)
            {
                if (word_chars[i] == word_chars[j])
                {
                    count++;
                }
            }
        }

        return count;
    }
};

class Solution2
{
public:
    int similarPairs(vector<string> &words)
    {
        int count = 0;
        map<set<char>, int> word_sets;

        for (const string &word : words)
        {
            set<char> char_set(word.begin(), word.end());
            word_sets[char_set]++;
        }

        for (const auto &pair : word_sets)
        {
            int n = pair.second;
            count += (n * (n - 1)) / 2;
        }

        return count;
    }
};

class Solution3
{
public:
    int similarPairs(vector<string> &words)
    {
        int count = 0;
        unordered_map<int, int> word_masks;

        for (const string &word : words)
        {
            int mask = 0;
            for (char chr : word)
            {
                mask |= (1 << (chr - 'a'));
            }
            word_masks[mask]++;
        }

        for (const auto &pair : word_masks)
        {
            int n = pair.second;
            count += (n * (n - 1)) / 2;
        }

        return count;
    }
};

int main()
{
    vector<string> words = {"abc", "bca", "cab", "xyz", "zyx"};
    Solution sol;
    int result = sol.similarPairs(words);
    std::cout << result << std::endl;
    return 0;
}